Sources/PlaydateKit/Core/Sound.swift

```diff
@@ -1,5 +1,113 @@
+        /// Returns 1 on success, 0 if buffer allocation failed.
+        @discardableResult
+        public func play(repeat: Int32 = 1) -> Int32 {
+            fileplayer.play(pointer, `repeat`)
+        }
+
+        /// Pauses the file player.
+        public func pause() {
+            fileplayer.pause(pointer)
+        }
+
+        /// Stops playing the file.
+        public func stop() {
+            fileplayer.stop(pointer)
+        }
+
+        /// Sets the buffer length of player to bufferLen seconds;
+        func setBufferLength(bufferLength: Float) {
+            fileplayer.setBufferLength(pointer, bufferLength)
+        }
+
+        /// Returns the length, in seconds, of the file loaded into player.
+        func getLength() -> Float {
+            fileplayer.getLength(pointer)
+        }
+
+        /// Returns `true` if player has underrun, `false` if not.
+        func didUnderrun() -> Bool {
+            fileplayer.didUnderrun(pointer) == 1
+        }
+
+        /// Sets the start and end of the loop region for playback, in seconds.
+        func setLoopRange(start: Float, end: Float) {
+            fileplayer.setLoopRange(pointer, start, end)
+        }
+
+        /// Sets the current offset in seconds.
+        func setOffset(_ offset: Float) {
+            fileplayer.setOffset(pointer, offset)
+        }
+
+        /// Returns the current offset in seconds for player.
+        func getOffset() -> Float {
+            fileplayer.getOffset(pointer)
+        }
+
+        /// Sets the playback rate for the player. 1.0 is normal speed, 0.5 is down an octave, 2.0 is up an octave, etc.
+        /// Unlike sampleplayers, fileplayers can’t play in reverse (i.e., rate < 0).
+        func setRate(_ rate: Float) {
+            fileplayer.setRate(pointer, rate)
+        }
+
+        /// Returns the playback rate for player.
+        func getRate() -> Float {
+            fileplayer.getRate(pointer)
+        }
+
+        /// If `flag` evaluates to true, the player will restart playback (after an audible stutter) as soon as data is available.
+        func setStopOnUnderrun(flag: Bool) {
+            fileplayer.setStopOnUnderrun(pointer, flag ? 1 : 0)
+        }
+
+        /// Sets the playback volume for left and right channels of player.
+        func setVolume(left: Float, right: Float) {
+            fileplayer.setVolume(pointer, left, right)
+        }
+
+        /// Sets the playback volume of player.
+        func setVolume(_ volume: Float) {
+            fileplayer.setVolume(pointer, volume, volume)
+        }
+
+        /// Gets the left and right channel playback volume for player.
+        func getVolume() -> (Float, Float) {
+            var left: Float = 0
+            var right: Float = 0
+            fileplayer.getVolume(pointer, &left, &right)
+            return (left, right)
+        }
+
+        /// Frees the given player.
+        public func free() {
+            _ =  fileplayer.freePlayer(pointer)
```

Comment:
The freePlayer call should be moved to the FilePlayer deinit, similar to `Video.Player` in `Graphics.swift`. Then you don't need to worry about memory management and can't get into a state where you have leaked memory.

Reply:
Good hint … did spot this pattern but forgot about it. Added this change. 👍